brokers/artifacts/broker.go

```diff
@@ -0,0 +1,179 @@
+//
+// Copyright (c) 2019 Red Hat, Inc.
+// This program and the accompanying materials are made
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package artifacts
+
+import (
+	"fmt"
+	"path/filepath"
+
+	jsonrpc "github.com/eclipse/che-go-jsonrpc"
+	"github.com/eclipse/che-plugin-broker/common"
+	"github.com/eclipse/che-plugin-broker/model"
+	"github.com/eclipse/che-plugin-broker/utils"
+)
+
+const errorNoExtFieldsTemplate = "Field 'extensions' is not found in the description of the plugin '%s'"
+
+// Broker is used to process Che plugins
+type Broker struct {
+	common.Broker
+	ioUtils utils.IoUtil
+	rand    common.Random
+}
+
+// NewBroker creates Che broker instance
+func NewBroker(localhostSidecar bool) *Broker {
+	return &Broker{
+		Broker:  common.NewBroker(),
+		ioUtils: utils.New(),
+		rand:    common.NewRand(),
+	}
+}
+
+func (b *Broker) fail(err error) error {
+	b.PubFailed(err.Error())
+	b.PubLog(err.Error())
+	return err
+}
+
+// PushEvents sets given tunnel as consumer of broker events.
+func (b *Broker) PushEvents(tun *jsonrpc.Tunnel) {
+	b.Broker.PushEvents(tun, model.BrokerStatusEventType, model.BrokerResultEventType, model.BrokerLogEventType)
+}
+
+// Start downloads metas from plugin registry for specified
+// pluginFQNs and then executes plugins metas processing and sending data to Che master
+func (b *Broker) Start(pluginFQNs []model.PluginFQN, defaultRegistry string) error {
+	b.cleanupPluginsDirectory(b.ioUtils)
+
+	pluginMetas, err := utils.GetPluginMetas(pluginFQNs, defaultRegistry, b.ioUtils)
+	if err != nil {
+		return b.fail(fmt.Errorf("Failed to download plugin meta: %s", err))
+	}
+	defer b.CloseConsumers()
+	b.PubStarted()
+	b.PrintInfo("Downloading plugin extensions")
+	b.PrintPlan(pluginMetas)
```

Comment:
Artifacts plugin broker prints the same output that metadata broker does, maybe it makes sense to omit it or print artifacts specific info like
```
List of plugins artifacts need to be downloaded
- redhat-developer/che-omnisharp-plugin/latest - This plug-in registers omnisharp-roslyn as a language server and provides project dependencies and language syntax.
- redhat-developer/netcoredbg-theia-plugin/latest - This plug-in provides Samsung/netcoredbg which implements VSCode Debug Adapter protocol and allows to debug .NET apps under .NET Core runtime.
- sonarsource/sonarlint-vscode/latest - VS Code extension that provides sonarlint features
```